### PR TITLE
Maintenance/511 requirement levels

### PR DESCRIPTION
**Pull request by CDR-API-Stream**
Consistency update for the treatment of requirement level keywords

### CODE REVIEW

File: slate/source/includes/security/endpoints/_register.md [hunk 1]
```diff
@@ -40,11 +40,11 @@ Participants will be required to register base URIs against each of their brands
 
 | Base URI | DH Brand | ADR Brand | Description
 |-----------|------|------|-----------------------------------------------------------------------------------------------|
-|**PublicBaseUri**|	<i class="icon-check"></i> | | Base URI for the Consumer Data Standard public endpoints. This should encompass all endpoints not requiring authentication |
-|**ResourceBaseUri**|	<i class="icon-check"></i> | | Base URI for the Consumer Data Standard resource endpoints. This should encompass all CDS resource endpoints requiring authentication |
+|**PublicBaseUri**|	<i class="icon-check"></i> | | Base URI for the Consumer Data Standard public endpoints. This **SHOULD** encompass all endpoints not requiring authentication |
+|**ResourceBaseUri**|	<i class="icon-check"></i> | | Base URI for the Consumer Data Standard resource endpoints. This **SHOULD** encompass all CDS resource endpoints requiring authentication |
```
Comment (perlboy) [inline slate/source/includes/security/endpoints/_register.md:44]: These are likely MUSTs as, if they weren't implemented things would break.

```suggestion
|**PublicBaseUri**|	<i class="icon-check"></i> | | Base URI for the Consumer Data Standard public endpoints. This **MUST** encompass all endpoints not requiring authentication |
|**ResourceBaseUri**|	<i class="icon-check"></i> | | Base URI for the Consumer Data Standard resource endpoints. This **MUST** encompass all CDS resource endpoints requiring authentication |
```

File: slate/source/includes/security/endpoints/_oidc_provider_configuration.md [hunk 2]
```diff
@@ -61,7 +61,7 @@ At a minimum, the Data Holder metadata MUST include:
 - `claims_supported`:  The list of supported claims.
 - `acr_values_supported`:  The supported ACR values.
 - `jwks_uri`: The JSON Web Key Set for the data holder.
-- `id_token_encryption_alg_values_supported`: The list of the supported JWE algorithms for securing the issued ID tokens. Must conform to **[[FAPI-RW-Draft]](#nref-FAPI-RW-Draft)** and **[[OIDD]](#nref-OIDD)**.
+- `id_token_encryption_alg_values_supported`: The list of the supported JWE algorithms for securing the issued ID tokens. **MUST** conform to **[[FAPI-RW-Draft]](#nref-FAPI-RW-Draft)** and **[[OIDD]](#nref-OIDD)**.
```
Comment (perlboy) [inline slate/source/includes/security/endpoints/_oidc_provider_configuration.md:64]: FAPI-RW (should be 1.0 btw) is a profile over the top so it is not necessary to do a MUST binding to OIDD.

```suggestion
- `id_token_encryption_alg_values_supported`: The list of the supported JWE algorithms for securing the issued ID tokens. **MUST** conform to **[[FAPI-RW-Draft]](#nref-FAPI-RW-Draft)**.
```

File: slate/source/includes/security/endpoints/_register.md [hunk 3]
```diff
@@ -40,11 +40,11 @@ Participants will be required to register base URIs against each of their brands
...
 |**InfoSecBaseUri**|	<i class="icon-check"></i> | | Base URI for the Consumer Data Standard InfoSec endpoints. This provides ADRs reference to the [OIDC Discovery Endpoint](https://openid.net/specs/openid-connect-discovery-1_0.html) |
```
Comment (perlboy) [inline slate/source/includes/security/endpoints/_register.md:45]:
```suggestion
|**InfoSecBaseUri**|	<i class="icon-check"></i> | | Base URI for the Consumer Data Standard InfoSec endpoints. This **MUST** be present and match the `iss` exposed at the [OIDC Discovery Endpoint](https://openid.net/specs/openid-connect-discovery-1_0.html) |
```

File: slate/source/includes/security/endpoints/_oidc_provider_configuration.md [hunk 4]
```diff
@@ -46,9 +46,9 @@ Content-Type: application/json
 | Client Authentication Required| No|
 | Bearer Token Required| No|
 
-Data Holders MUST make their OpenID Provider Metadata available via a configuration end point as outlined in [Section 3 and 4 of the OpenID Connect Discovery standards](https://openid.net/specs/openid-connect-discovery-1_0.html) **[[OIDD]](#nref-OIDD)**.
+Data Holders **MUST** make their OpenID Provider Metadata available via a configuration end point as outlined in [Section 3 and 4 of the OpenID Connect Discovery standards](https://openid.net/specs/openid-connect-discovery-1_0.html) **[[OIDD]](#nref-OIDD)**.
```
Comment (perlboy) [inline slate/source/includes/security/endpoints/_oidc_provider_configuration.md:49]: This is a duplicate of `InfoSecBaseURI` and in fact the metadata url is calculated from the `iss`

```suggestion
```

File: slate/source/includes/security/_participant_statuses.md.erb [hunk 5]
```diff
@@ -75,7 +75,7 @@ The CDR Registrar has the ability to change the status of a Software Product ind
...
 * The CDR Register **MUST NOT** provide an undefined status for Data Recipient or Data Recipient Software Product Statuses
-* If Data Holders do not receive a status for a Data Recipient or Data Recipient Software Product, or receives a status that is not recognised, Data Holders **SHOULD** ignore the value and use the previous status value retrieved from the CDR Register. Data Holders **SHOULD** continue to use the previous status value until a valid value is returned by the CDR Register or the ACCC informs the Data Holder using an alternative mechanism. There is **no** upper bound for how long previous status values should remain trusted.
+* If Data Holders do not receive a status for a Data Recipient or Data Recipient Software Product, or receives a status that is not recognised, Data Holders **SHOULD** ignore the value and use the previous status value retrieved from the CDR Register. Data Holders **SHOULD** continue to use the previous status value until a valid value is returned by the CDR Register or the ACCC informs the Data Holder using an alternative mechanism. There is **no** upper bound for how long previous status values **SHOULD** remain trusted.
```
Comment (perlboy) [inline slate/source/includes/security/_participant_statuses.md.erb:78]: This doesn't consider problems associated with communicated statuses for DR not aligning with SP

File: slate/source/includes/security/_consent.md [hunk 6]
```diff
@@ -27,15 +27,15 @@ The following [normal](https://openid.net/specs/openid-connect-core-1_0.html#Nor
...
+- `phone_number_verified`: True if the End-User's phone number has been verified; otherwise false. When this Claim Value is true, this means that the Data Holder took affirmative steps to ensure that this phone number was controlled by the End-User at the time the verification was performed. The means by which a phone number is verified is context- specific, and dependent upon the trust framework or contractual agreements within which the parties are operating. When true, the phone_number Claim **MUST** be in **[[E.164]](#iref-E-164)** format and any extensions **MUST** be represented in **[[RFC3966]](#iref-RFC3966)** format.
```
Comment (perlboy) [inline slate/source/includes/security/_consent.md:38]: Restating the OIDC spec is overkill imo.

File: slate/source/includes/security/_consent.md [hunk 7]
```diff
@@ -27,15 +27,15 @@ The following [normal](https://openid.net/specs/openid-connect-core-1_0.html#Nor
...
+- `refresh_token_expires_at`: indicates the date-time at which the most recently provided refresh token will expire. Its value **MUST** be a number containing a NumericDate value, as specified in section 2 of [section 2](https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-32#section-2) **[[JWT]](#nref-JWT)**.  If no refresh token has been provided then a zero value **SHOULD** be returned.
+- `sharing_expires_at`: indicates the date-time at which the current sharing arrangement will expire. Its value **MUST** be a number containing a NumericDate value, as specified in [section 2](https://tools.ietf.org/html/draft-ietf-oauth-json-web-token-32#section-2) of **[[JWT]](#nref-JWT)**.  If consent is not complete or a `sharing_duration` was not requested in the authorisation request object then a zero value **SHOULD** be returned.
 
 The following standard **[[OIDC]](#nref-OIDC)** claims **MAY** be supported:
```
Comment (perlboy) [inline slate/source/includes/security/_consent.md:33]:
```suggestion
The **[[OIDC]](#nref-OIDC)** claims `email`, `email_verified`, `phone_number`, `phone_number_verified` **MAY** be supported.
```

File: slate/source/includes/security/_certificate_management.md [hunk 8]
```diff
@@ -50,7 +50,7 @@ Please refer to the [Register onboarding guide](https://www.accc.gov.au/focus-ar
 Further details on the Register CA issued certificates can be found in the [ACCC Certificate Practice Statement V1.0](https://www.cdr.gov.au/sites/default/files/2020-12/CDR%20-%20ACCC%20Certification%20practice%20statement.pdf).
 
 ### Certificate Validation
-Certificate validation must check:
+Certificate validation **MUST** check:
```
Comment (perlboy) [inline slate/source/includes/security/_certificate_management.md:53]: Very bad idea to capitalise this with some random bullet points. Suggest rewording this section to refer to RFCs or not make it 2119.

File: slate/source/includes/secondary/_energy_security.md [hunk 9]
```diff
@@ -7,7 +7,7 @@ For the energy sector requests for the following data clusters are Shared Respon
 * Energy Usage
 * Distributed Energy Resources
 
-CDR requests for these data clusters to a Data Holder MUST be fulfilled by a request to AEMO.
+CDR requests for these data clusters to a Data Holder **MUST** be fulfilled by a request to AEMO.
```
Comment (perlboy) [inline slate/source/includes/secondary/_energy_security.md:10]: Not related but here's a suggestion to allow for usage to be provided voluntarily by non AEMO sources.

```suggestion
CDR requests for these data clusters to a Data Holder **MUST** be fulfilled by a request to AEMO where AEMO is authoritative for that data.
```

File: slate/source/includes/secondary/_energy_apis.md [hunk 10]
```diff
@@ -27,28 +27,28 @@ would become the secondary endpoint:
...
+* General headers **SHOULD** be provided as if the request were coming from the primary Data Holder and not propagated from the call made by the Data Recipient
+* The `x-fapi-interaction-id` header **MUST** be propagated from the Data Recipient call to AEMO to allow for end to end tracing. If not supplied by the Data Recipient, the primary Data Holder **MUST** create a unique value for the `x-fapi-interaction-id` header before calling AEMO
 * Endpoints that require knowledge of the NMIs that belong to the CDR Consumer have been
-excluded from the AEMO endpoint set. This includes *Get Bulk Usage* and *Get Bulk DER*.  When a primary Data Holder is required to respond to these endpoints they should call the equivalent
+excluded from the AEMO endpoint set. This includes *Get Bulk Usage* and *Get Bulk DER*.  When a primary Data Holder is required to respond to these endpoints they **SHOULD** call the equivalent
 endpoint for specific service points and provide the specific list of NMIs to AEMO
```
Comment (perlboy) [inline slate/source/includes/secondary/_energy_apis.md:47]: Isn't this a MUST? There's no other way and it opens the door to calling individual NMIs?

```suggestion
excluded from the AEMO endpoint set. This includes *Get Bulk Usage* and *Get Bulk DER*.  When a primary Data Holder is required to respond to these endpoints they **MUST** call the *Get Usage For Specific Service Points (SR)* and  *Get DER For Specific Service Points (SR)* respectively.
```

File: slate/source/includes/secondary/_energy_apis.md [hunk 11]
```diff
@@ -27,28 +27,28 @@ would become the secondary endpoint:
...
-* All occurrences of the `servicePointId` field, whether in a request, a response, or as an input parameter (such as path parameter or query parameter) should be populated with the equivalent `NationalMeteringId` in plain text
-* Fields in the links object for all responses MUST be translated by the Data Holder into
+* The `x-fapi-auth-date` header **MUST NOT** be passed to AEMO and AEMO **MUST NOT** require this header
+* The `x-fapi-customer-ip-address` header **MUST NOT** be passed to AEMO and AEMO **MUST NOT** require this header
+* The `x-cds-client-headers` header **MUST NOT** be passed to AEMO and AEMO **MUST NOT** require this header
+* A new header named `x-cds-arrangement` **MUST** be passed to AEMO for every invocation. This header **SHOULD** contain the arrangement ID for the consent that the request is being made under and will be used for tracing and audit purposes. This field **MUST** be populated but AEMO **MUST NOT** seek to validate the consent associated with the arrangement
```
Comment (perlboy) [inline slate/source/includes/secondary/_energy_apis.md:33]: Restating the same thing is bad form.

```suggestion
* A new header named `x-cds-arrangement` **MUST** be passed to AEMO for every invocation. This header **SHOULD** contain the arrangement ID for the consent that the request is being made under and will be used for tracing and audit purposes. AEMO **MUST NOT** seek to validate the consent associated with the arrangement
```

File: slate/source/includes/register/_discovery_apis.md [hunk 12]
```diff
@@ -681,7 +681,7 @@ This operation does not require authentication
 |alg|string|true|none|The "alg" (algorithm) parameter identifies the algorithm intended for use with the key|
 |e|string|true|none|The "e" RSA public exponent parameter|
 |key_ops|[string]|true|none|The "key_ops" (key operations) parameter identifies the operation(s) for which the key is intended to be used|
-|kid|string|true|none|The "kid" (key ID) parameter is partially used to match a specific key. Note the "kid" parameter is not guaranteed unique and additional parameters should be used to progressively to identify a key within a set|
+|kid|string|true|none|The "kid" (key ID) parameter is partially used to match a specific key. Note the "kid" parameter is not guaranteed unique and additional parameters **SHOULD** be used to progressively to identify a key within a set|
```
Comment (perlboy) [inline slate/source/includes/register/_discovery_apis.md:684]: Other parts of the Standard mandate it to be unique.....

File: slate/source/includes/register/_discovery_apis.md [hunk 13]
```diff
@@ -498,7 +498,7 @@ Endpoint used by participants to discover data recipients and associated brands
 |Name|In|Type|Required|Description|
 |---|---|---|---|---|
 |industry|path|string|true|The industry the participant is retrieving data for (Banking, etc)|
-|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client. Must be set to a positive integer.|
+|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client. **MUST** be set to a positive integer.|
```
Comment (perlboy) [inline slate/source/includes/register/_discovery_apis.md:501]: Restating previous statements.

```suggestion
|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client.|
```

File: slate/source/includes/register/_discovery_apis.md [hunk 14]
```diff
@@ -428,7 +428,7 @@ Endpoint used by participants to discover the statuses for Data Recipients from
...
```
Comment (perlboy) [inline slate/source/includes/register/_discovery_apis.md:431]: Restating previous statements.

```suggestion
|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client.|
```

File: slate/source/includes/register/_discovery_apis.md [hunk 15]
```diff
@@ -358,7 +358,7 @@ Endpoint used by participants to discover the statuses for software products fro
...
```
Comment (perlboy) [inline slate/source/includes/register/_discovery_apis.md:361]: Restating previous statements.

```suggestion
|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client.|
```

File: slate/source/includes/register/_discovery_apis.md [hunk 16]
```diff
@@ -291,7 +291,7 @@ Get a Software Statement Assertion (SSA) for a software product on the CDR Regis
 |dataRecipientBrandId|path|string|true|Unique id for the Accredited Data Recipient Brand that the Software Product is associated with in the CDR Register. Refer to [Identifiers](https://cdr-register.github.io/register/#identifiers) for details|
 |softwareProductId|path|string|true|Unique id for the Accredited Data Recipient Software Product in the CDR Register. Refer to [Identifiers](https://cdr-register.github.io/register/#identifiers) for details|
 |Authorization|header|string|true|An Authorisation Token as per [RFC6750](https://tools.ietf.org/html/rfc6750)|
-|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client. Must be set to a positive integer.|
+|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client. **MUST** be set to a positive integer.|
```
Comment (perlboy) [inline slate/source/includes/register/_discovery_apis.md:294]: Restating previous statements.

```suggestion
|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client.|
```

File: slate/source/includes/register/_discovery_apis.md [hunk 17]
```diff
@@ -254,7 +254,7 @@ Allows Data Recipients to discover data holder brands available in the CDR ecosy
 
 
 <aside class="notice">
-To perform this operation, the Data Recipient or Data Holder must be authenticated and authorised with the following scope: [cdr-register:bank:read](#authorisation-scopes).
+To perform this operation, the Data Recipient or Data Holder **MUST** be authenticated and authorised with the following scope: [cdr-register:bank:read](#authorisation-scopes).
```
Comment (perlboy) [inline slate/source/includes/register/_discovery_apis.md:257]: scopes are not authenticated they are only authorised.

```suggestion
To perform this operation, the Data Recipient or Data Holder **MUST** be authorised with the following scope: [cdr-register:bank:read](#authorisation-scopes).
```

File: slate/source/includes/register/_discovery_apis.md [hunk 18]
```diff
@@ -172,7 +172,7 @@ Allows Data Recipients to discover data holder brands available in the CDR ecosy
 |page|query|integer(int32)|false|the page number to return|
 |page-size|query|integer(int32)|false|the number of records to return per page|
 |Authorization|header|string|true|An Authorisation Token as per [RFC6750](https://tools.ietf.org/html/rfc6750)|
-|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client. Must be set to a positive integer.|
+|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client. **MUST** be set to a positive integer.|
```
Comment (perlboy) [inline slate/source/includes/register/_discovery_apis.md:175]: Restating previous statements.

```suggestion
|x-v|header|string|false|The [version](https://consumerdatastandardsaustralia.github.io/standards/#response-headers) of the API end point requested by the client.|
```

File: slate/source/includes/register/_dcr_apis.md [hunk 19]
```diff
@@ -416,14 +416,14 @@ To perform this operation, you must be authenticated and authorised with the fol
...
 |tos_uri|string|false|none|URL string that points to a human-readable terms of service document for the Software Product|
 |policy_uri|string|false|none|URL string that points to a human-readable policy document for the Software Product|
 |jwks_uri|string|true|none|URL string referencing the client JSON Web Key (JWK) Set [RFC7517] document, which contains the client public keys|
 |revocation_uri|string|false|none|URI string that references the location of the Software Product consent revocation endpoint|
-|recipient_base_uri|string|false|none|Base URI for the Consumer Data Standard Data Recipient endpoints. This should be the base to provide reference to all other [Data Recipient Endpoints](https://consumerdatastandardsaustralia.github.io/standards/#end-points)|
+|recipient_base_uri|string|false|none|Base URI for the Consumer Data Standard Data Recipient endpoints. This **SHOULD** be the base to provide reference to all other [Data Recipient Endpoints](https://consumerdatastandardsaustralia.github.io/standards/#end-points)|
```
Comment (perlboy) [inline slate/source/includes/register/_dcr_apis.md:426]: Suggested elsewhere, search for that to find commentary.

```suggestion
|recipient_base_uri|string|false|none|Base URI for the Consumer Data Standard Data Recipient endpoints. This **MUST** be the base to provide reference to all other [Data Recipient Endpoints](https://consumerdatastandardsaustralia.github.io/standards/#end-points)|
```

File: slate/source/includes/nfrs/_session-requirements.md [hunk 20]
```diff
@@ -1,4 +1,4 @@
 ## Session Requirements
 A expiry time of a unique session should be set according to the statements included in the [Security Profile](#security-profile).
 
-After a unique session is expired it is expected that the Data Recipient Software Product, for the same customer, may establish a new session as long as the authorisation is still valid.
+After a unique session is expired it is expected that the Data Recipient Software Product, for the same customer, **MAY** establish a new session as long as the authorisation is still valid.
```
Comment (perlboy) [inline slate/source/includes/nfrs/_session-requirements.md:4]:
```suggestion
After a unique session is expired it is expected that the Data Recipient Software Product, for the same Arrangement, **SHOULD** establish a new session as long as the authorisation is still valid.
```

File: slate/source/includes/nfrs/_service-exemptions.md [hunk 21]
```diff
@@ -2,6 +2,6 @@
...
 
-- Periods of time when the digital channels for the data holder are the target for a distributed denial of service or equivalent form of attack (this should result in http error `429 Too Many Requests` being returned).
-- A significant increase in traffic from a poorly designed or misbehaving Data Recipient Software Product (this should result in http error `429 Too Many Requests` being returned).
-- If the data holder identifies a situation where there is the potential for physical or financial harm or abuse (this should result in http error `403 Forbidden` being returned).
+- Periods of time when the digital channels for the data holder are the target for a distributed denial of service or equivalent form of attack (this **SHOULD** result in http error `429 Too Many Requests` being returned).
+- A significant increase in traffic from a poorly designed or misbehaving Data Recipient Software Product (this **SHOULD** result in http error `429 Too Many Requests` being returned).
+- If the data holder identifies a situation where there is the potential for physical or financial harm or abuse (this **SHOULD** result in http error `403 Forbidden` being returned).
```
Comment (perlboy) [inline slate/source/includes/nfrs/_service-exemptions.md:7]: There's no error payload for 403 that is suitable here. We currently throw Error/Expected.....

**Review by perlboy — COMMENTED** [review]
There seems to be non 2119 terms used which is not appropriate. Additionally there seems to be excessive 2119 language in use. While this may be a first draft what is proposed makes significant changes to the underlying Standards as they are understood at present. I've walked through as much as I can to provide assistance but this needs a lot more work.

I think it's appropriate to quote 2119 here as the approach taken by the DSB seems to be one of mass search and replace:

> Imperatives of the type defined in this memo must be used with care and sparingly.  In particular, they MUST only be used where it is actually required for interoperation or to limit behavior which has potential for causing harm (e.g., limiting retransmisssions)  For example, they must not be used to try to impose a particular method on implementors where the method is not required for interoperability.

It's a pretty big line ball as to whether many of the 2119 statements proposed _actually_ meet this requirement or are simply commentary dressed up as Standards.

**Comment by CDR-API-Stream** [comment]
This PR won't be included in 1.18.0. Based on @perlboy's feedback it makes sense to address the consistency and use of the requirement level keywords.
